templates/node/node.py

```diff
@@ -0,0 +1,269 @@
+
+    VDC_TEMPLATE = 'github.com/openvcloud/0-templates/vdc/0.0.1'
+    SSH_TEMPLATE = 'github.com/openvcloud/0-templates/sshkey/0.0.1'
+
+    def __init__(self, name, guid=None, data=None):
+        super().__init__(name=name, guid=guid, data=data)
+
+        self._ovc = None
+        self._vdc = None
+        self._machine = None
+
+    def validate(self):
+        if not self.data['vdc']:
+            raise RuntimeError('vdc name should be given')
+        # ensure uploaded key
+        self.sshkey
+
+    def update_data(self, data):
+        # merge the new data
+        self.data.update(data)
+        self.save()
+
+    @property
+    def ovc(self):
+        """
+        An ovc connection instance
+        """
+        if self._ovc is not None:
+            return self._ovc
+        
+        vdc = self.vdc
+        instance = vdc.ovc.instance
+        self._ovc = j.clients.openvcloud.get(instance=instance)
+
+        return self._ovc
+
+    @property
+    def vdc(self):
+        if self._vdc:
+            return self._vdc
+
+        # Get object for an VDC service, make sure exactly one is running
+        vdc = self.api.services.find(template_uid=self.VDC_TEMPLATE, name=self.data['vdc'])
+        if len(vdc) != 1:
+            raise RuntimeError('found %s vdc, requires exactly 1' % len(vdc))
+        
+        self._vdc = vdc[0]
+
+        return self._vdc
+
+    @property
+    def space(self):
+        return self.ovc.space_get(
+            accountName=None,
+            spaceName=self.data['vdc']
+            )
+
+    @property
+    def sshkey(self):
+        """ Get a path and keyname of the sshkey service """
+
+        sshkeys = self.api.services.find(template_uid=self.SSH_TEMPLATE)
+        if not len(sshkeys):
+            raise RuntimeError('no %s ssh services found' % len(sshkeys))
+
+        # Get key name and path
+        path = sshkeys[0].data['path']
+        key = path.split('/')[-1]
+
+        return key         
+
+    @property
+    def machine(self):
+        if self._machine:
+            return self._machine
+        return self.space.machines.get(self.name)
+
+    def install(self):
+        machine = self.machine
+        if not machine:
+            machine = self._machine_create()
+
+        # Get data from the vm
+        ip_private, vm_info = machine.machineip_get()
+        self.data['sshLogin'] = vm_info['accounts'][0]['login']
+        self.data['sshPassword']= vm_info['accounts'][0]['password']
+        self.data['ipPrivate'] = ip_private
+        self.data['ipPublic'] = machine.space.model['publicipaddress']
+        self.data['machineId'] = machine.id
+
+        self.portforward_create(self.data['ports'])
+        # TODO: self._configure_disks()
+
+        self.state.set('actions', 'install', 'ok')
+        self.save()
+
+    def uninstall(self):
+        # check if the machine is in the space
+        machine = self.machine
+        if machine:
+            self.machine.delete()
+        else:
```

Comment:
No need to raise when you try to delete something and this something is already gone